Remove logic to check if a named arg could be an assignment (and enable `-Xsource:2.13` by default)

lrytz opened the pull request:
An `x = e` expression in parameter position is now a named argument,
syntactically, unconditionally.

Fixes scala/scala-dev#426

lrytz committed:
Deprecate assignments in argument position

Issue a deprecation warning when an assignment expression (`x = e`)
in a parameter position is treated as an assignment, not a named
argument.

In 2.13 this will be an error, `f(x = e)` will be restricted to
named arguments and never be treated as an assignment to `x`.

The 2.13 behavior is available under `-Xsource:2.13`

See scala/scala-dev#426

lrytz committed:
Rename AssignOrNamedArg to NamedArg

lrytz committed:
Remove logic to check if a named arg could be an assignment

An `x = e` expression in parameter position is now a named argument,
syntactically, unconditionally.

Fixes scala/scala-dev#426

lrytz committed:
Change default Xsource to 2.13

Adjust tests as necessary.

Remove run test for t10097, it produces a compiler error under
Xsource:2.13. This is tested in a neg test.

lrytz commented:
Hehe, our current 2.13.x branch doesn't compile with `-Xsource:2.13`
```
[error] /home/jenkins/workspace/scala-2.13.x-validate-test/src/library/scala/sys/process/BasicIO.scala:165: type mismatch;
[error]  found   : String
[error]  required: () => String
[error]     try processLinesFully(processLine)(reader.readLine)
[error]                                               ^
```

adriaanm commented on src/compiler/scala/tools/nsc/settings/ScalaSettings.scala:
```diff
@@ -133,7 +135,7 @@ trait ScalaSettings extends AbsScalaSettings
   val reporter           = StringSetting       ("-Xreporter", "classname", "Specify a custom reporter for compiler messages.", "scala.tools.nsc.reporters.ConsoleReporter")
   val strictInference    = BooleanSetting      ("-Xstrict-inference", "Don't infer known-unsound types")
-  val source             = ScalaVersionSetting ("-Xsource", "version", "Treat compiler input as Scala source for the specified version, see scala/bug#8126.", initial = ScalaVersion("2.12"))
+  val source             = ScalaVersionSetting ("-Xsource", "version", "Treat compiler input as Scala source for the specified version, see scala/bug#8126.", initial = ScalaVersion("2.13"))
```
good catch!

adriaanm commented on src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala:
```diff
@@ -46,9 +46,10 @@ trait ContextErrors {
+   * assignment (to figure out if the expression was a valid assignment, a4a892fb01). Since 2.13 we
+   * could use a normal NormalTypeError, but I'll leave it in for now, it might come in handy.
    */
   class NormalTypeErrorFromCyclicReference(underlyingTree: Tree, errMsg: String)
```
I'm always weary of leaving unused stuff in, because some future dev may misunderstand its purpose and wreak havoc with it :-)

lrytz replied:
OK, I'm happy to remove it.

lrytz committed:
Make sure `@elidable` is only on concrete methods

Remove `@elidable` from abstract methods in tests, improve the error
message. Also improve the scaladoc on `@elidable`.

lrytz commented on test/files/run/t7187.scala:
```diff
@@ -0,0 +1,6 @@
+object Test extends App {
+  def foo(): () => String = () => "hi there"
+  val f: () => Any = foo
```
Note that in 2.13 (`-Xsource:2.13`) this compiles without a warning, `foo` is applied to the empty parameter list.

In 2.12, eta-expansion comes in before, and it compiles with a deprecation warning:
```
➜  sandbox git:(syntacticNamedArgs-213) scalac Test.scala -deprecation
Test.scala:3: warning: Eta-expansion of zero-argument method values is deprecated. Did you intend to write Test.this.foo()?
  val f: () => Any = foo
                     ^
one warning found
```

So the program compiles in both versions, but differently.

lrytz committed:
Changes to allow compilation with -Xsource:2.13

Nullary methods are no longer eta-expanded.
Assignments in paramter position need to be wrapped in brackets.

adriaanm commented:
Sorry for the delay, Lukas.

lrytz commented:
This merge broke the build, as noted in https://github.com/scala/scala/pull/6143#issuecomment-348407519

These are most likely the relevant commits that went into 2.13.x after this PR's parent:
  - https://github.com/scala/scala/pull/5983
  - https://github.com/scala/scala/pull/6069
